include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,1724 @@
+   * of the quadrature points is known to match a priori. An assertion is thrown
+   * if the quadrature points and weights of the two FEValuesBase objects do not
+   * match one-to-one.
+   */
+  matching,
+
+  /**
+   * The FEValuesBase objects provide the same quadrature rule over the same
+   * set, but with possibly different orientations. The quadrature points are
+   * reordered internally, so that the resulting quadrature points and weights
+   * are the same as in the matching case. An assertion is thrown if the
+   * quadrature points and weights of the two FEValuesBase cannot be reordered
+   * to match one-to-one.
+   */
+  reorder,
+
+  /**
+   * The FEValuesBase objects provide partially overlapping quadrature rules
+   * over two intersecting sets, with possibly different orientations. The
+   * quadrature points are reordered and matched internally, so that the
+   * resulting quadrature points and weights are only the ones that match
+   * between the two objects. If no overlapping points and weights can be found,
+   * an empty set is used to integrate, and the resulting integral is zero. No
+   * assertion is thrown in this case.
+   */
+  overlapping,
+};
+
+/**
+ * DoF coupling options when assembling double integrals.
+ *
+ * When computing the approximation of double integrals of the form
+ *
+ * \f[
+ * \int_{T_1} \int{T_2} K(x_1, x_2) v_i(x_1) w_j(x_2) dT_1 dT_2,
+ * \f]
+ *
+ * where $T_1$ and $T_2$ are two arbitrary sets (cells, faces, edges, or any
+ * combination thereof), and $K$ is a (possibly singular) coupling kernel, one
+ * may want to combine degrees from two different FEValuesBase objects (i.e.,
+ * basis functions $v_i$ and $w_j$ in the examples above)
+ *
+ * This enum class provides a way to specify how the degrees of freedom should
+ * be combined. There are two cases of interest:
+ *
+ * 1. the two FEValuesBase objects refer to different DoFHandlers
+ * 2. the two FEValuesBase objects refer to the same DoFHandler
+ *
+ * In the first case, one usually treat the two sets of degrees of freedom as
+ * independent of each other, and the resulting matrix is generally rectangular.
+ *
+ * In the second case, one may choose to treat the two sets of degrees of
+ * freedom either as independent or to group them together. A similar approach
+ * is used in the FEInterfaceValues class, where the degrees of freedom of the
+ * two FEValuesBase objects are grouped together, in a contiguous way, so that
+ * the resulting basis functions are interpreted in the following way:
+ *
+ * \f[
+ * \phi_{l,i}(x) = \begin{cases} v_i(x) & \text{ if } i \in [0,n_l) \\
+ * 0 & \text{ if ) i \in [n_l, n_l+n_r] \end{cases},\quad \phi_{l,i}(x) =
+ * \begin{cases} 0(x) & \text{ if } i \in [0,n_l) \\
+ * w_{i-n_l} & \text{ if ) i \in [n_l, n_l+n_r] \end{cases},
+ * \f]
+ *
+ * where $phi_{l,i}$ is the left basis function with index $i$ and $n_{l,r}$ are
+ * the number of local dofs on the left and right FEValuesBase objects.
```

Comment:
Unfortunately there is no "interior" or "exterior" here (in principle). I don't like left and right, either. Maybe "first" and "second"?  Maybe something like 

```
fev[0][extractor]
fev[1][extractor]
```

could be better? No names, but an index. I think exterior and interior, or here and there do not make sense in this context, even though you could use this class for interfaces.

Reply:
I agree with this assessment, a name does not make much sense because it would be arbitrary. I would lean towards `first` and `second`, in analogy to `std::pair`, as opposed to the plain index, but the index would also be ok. Other opinions?